commons/src/main/java/com/jagrosh/jdautilities/commons/utils/InverseAction.java

```diff
@@ -0,0 +1,616 @@
+
+        return updated.getManager()
+                      .setAllow(allow).setDeny(deny);
+    }
+
+    /**
+     * @return An attempt to remove said emote
+     */
+    public static AuditableRestAction<Void> of(EmoteAddedEvent event)
+    {
+        return event.getEmote().delete();
+    }
+
+    /**
+     * @return An attempt to change the name back to what it just was
+     */
+    public static EmoteManager of(EmoteUpdateNameEvent event)
+    {
+        Emote emote = event.getEmote();
+        EmoteManager manager = emote.getManager();
+        String oldName = event.getOldName();
+
+        return manager.setName(oldName);
+    }
+
+    /**
+     * @return An attempt to change the roles to what they just were
+     */
+    public static EmoteManager of(EmoteUpdateRolesEvent event)
+    {
+        Emote emote = event.getEmote();
+        EmoteManager manager = emote.getManager();
+        HashSet<Role> oldRoles = new HashSet<>(event.getOldRoles());
+
+        return manager.setRoles(oldRoles);
+    }
+
+    /**
+     * @return An attempt to remove said role
+     */
+    public static AuditableRestAction<Void> of(RoleCreateEvent event)
+    {
+        return event.getRole().delete();
+    }
+
+    /**
+     * @return An attempt to add the role back
+     */
+    public static RoleAction of(RoleDeleteEvent event)
+    {
+        return event.getRole().createCopy();
+    }
+
+    /**
+     * @return An attempt to change the mentionable state back
+     */
+    public static RoleManager of(RoleUpdateMentionableEvent event)
+    {
+        Role role = event.getRole();
+        RoleManager manager = role.getManager();
+        Boolean oldState = event.getOldValue();
+
+        return manager.setMentionable(oldState);
+    }
+
+    /**
+     * @return An attempt to move the role back to where it just was
+     */
+    public static OrderAction<Role, RoleOrderAction> of(RoleUpdatePositionEvent event)
+    {
+        Role role = event.getRole();
+        int oldPos = event.getOldPosition();
+
+        return event.getGuild()
+                    .modifyRolePositions()
+                    .selectPosition(role)
+                    .moveTo(oldPos);
+    }
+
+    /**
+     * @return An attempt to change the role's permissions back
+     */
+    public static RoleManager of(RoleUpdatePermissionsEvent event)
+    {
+        Role role = event.getRole();
+        RoleManager manager = role.getManager();
+        EnumSet<Permission> oldPerms = event.getOldPermissions();
+
+        return manager.setPermissions(oldPerms);
+    }
+
+    /**
+     * @return An attempt to change the role's name back
+     */
+    public static RoleManager of(RoleUpdateNameEvent event)
+    {
+        Role role = event.getRole();
+        RoleManager manager = role.getManager();
+        String oldName = event.getOldName();
+
+        return manager.setName(oldName);
+    }
+
+    //TODO What the hell is a hoisted role? Lmao, look at that later
```

Comment:
basically this

![hoisted](https://the-negative-one-in.me/XBXwCSafg6.png)

Reply:
Thanks, man